File: smashpen/src/main/java/com/smashpen/smashpen/CharacterRestController.java
package com.smashpen.smashpen;

import com.smashpen.smashpen.repository.UserRepository;
import com.smashpen.smashpen.domain.SmashCharacter;
import com.smashpen.smashpen.domain.CharacterNotes;
import com.smashpen.smashpen.repository.CharacterNotesRepository;
import com.smashpen.smashpen.repository.CharacterRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import java.net.URI;
import java.util.Collection;

@CrossOrigin(origins = "http://localhost:3000")
@RestController
@RequestMapping("/{userId}/character")
public class CharacterRestController {

	private UserRepository userRepository;
	private CharacterNotesRepository characterNotesRepository;
	private CharacterRepository characterRepository;

	@Autowired
	CharacterRestController(UserRepository userRepository,
							CharacterNotesRepository characterNotesRepository,
							CharacterRepository characterRepository) {
		this.userRepository = userRepository;
		this.characterNotesRepository = characterNotesRepository;
		this.characterRepository = characterRepository;
	}

	@RequestMapping(method = RequestMethod.GET)
	Collection<CharacterNotes> readCharacters(@PathVariable Long userId) throws UserNotFoundException {
		this.validateUser(userId);
		return this.characterNotesRepository.findAllByUserId(userId);
	}

	@RequestMapping(method = RequestMethod.GET, value = "/{characterNotesId}")
	CharacterNotes readCharacter(@PathVariable Long userId, @PathVariable Long characterNotesId) throws UserNotFoundException {
		this.validateUser(userId);
		return this.characterNotesRepository.findOne(characterNotesId);
	}

	@RequestMapping(method = RequestMethod.POST)
	ResponseEntity<?> add(@PathVariable Long userId, @RequestBody CharacterNotes input) throws UserNotFoundException {
		this.validateUser(userId);
		return this.userRepository.findById(userId).map(account -> {
				SmashCharacter smashCharacter = characterRepository.findByName(input.getSmashCharacter().getName());
				CharacterNotes existingNotes = characterNotesRepository.findCharacterNotesBySmashCharacterAndUser(smashCharacter, account);
				CharacterNotes result;
				if (existingNotes != null) {
					existingNotes.setNotes(input.getNotes());
					result = characterNotesRepository.save(existingNotes);
				} else {
					result = characterNotesRepository.save(new CharacterNotes(account, smashCharacter, input.getNotes()));
				}
				URI location = ServletUriComponentsBuilder
							.fromCurrentRequest().path("/{id}")
							.buildAndExpand(result.getId()).toUri();
				return ResponseEntity.created(location).build();
			})
			.orElse(ResponseEntity.noContent().build());
	}


	private void validateUser(Long userId) throws UserNotFoundException {
		this.userRepository.findById(userId).orElseThrow(() -> new UserNotFoundException(userId));
		Long currentUserId = this.userRepository.findByUsername(SecurityContextHolder.getContext().getAuthentication().getName()).get().getId();
		if (!userId.equals(currentUserId)) {
			throw new SecurityException("Cannot access another users account");
		}
	}
}

File: smashpen/src/main/java/com/smashpen/smashpen/CharacterRestController.kt
package com.smashpen.smashpen

import com.smashpen.smashpen.repository.UserRepository
import com.smashpen.smashpen.domain.SmashCharacter
import com.smashpen.smashpen.domain.CharacterNotes
import com.smashpen.smashpen.repository.CharacterNotesRepository
import com.smashpen.smashpen.repository.CharacterRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.support.ServletUriComponentsBuilder

import java.net.URI

@CrossOrigin(origins = arrayOf("*"))
@RestController
@RequestMapping("/{userId}/character")
class CharacterRestController @Autowired
internal constructor(private val userRepository: UserRepository,
                     private val characterNotesRepository: CharacterNotesRepository,
                     private val characterRepository: CharacterRepository) {

    @RequestMapping(method = arrayOf(RequestMethod.GET))
    @Throws(UserNotFoundException::class)
    internal fun readCharacters(@PathVariable userId: Long?): Collection<CharacterNotes> {
        this.validateUser(userId)
        return this.characterNotesRepository.findAllByUserId(userId)
    }

    @RequestMapping(method = arrayOf(RequestMethod.GET), value = "/{characterNotesId}")
    @Throws(UserNotFoundException::class)
    internal fun readCharacter(@PathVariable userId: Long?, @PathVariable characterNotesId: Long?): CharacterNotes {
        this.validateUser(userId)
        return this.characterNotesRepository.findOne(characterNotesId)
    }

    @RequestMapping(method = arrayOf(RequestMethod.POST))
    @Throws(UserNotFoundException::class)
    internal fun add(@PathVariable userId: Long?, @RequestBody input: CharacterNotes): ResponseEntity<*> {
        this.validateUser(userId)
        return this.userRepository.findById(userId).map { account ->
            val smashCharacter = characterRepository.findByName(input.smashCharacter?.name)
            val existingNotes = characterNotesRepository.findCharacterNotesBySmashCharacterAndUser(smashCharacter, account)
            val result = if (existingNotes == null)
                characterNotesRepository.save(CharacterNotes(account, smashCharacter, input.notes))
            else
                characterNotesRepository.save(existingNotes)
            val location = ServletUriComponentsBuilder
                    .fromCurrentRequest().path("/{id}")
                    .buildAndExpand(result.id!!).toUri()
            ResponseEntity.created(location).build<Any>()
        }
                .orElse(ResponseEntity.noContent().build())
    }


    @Throws(UserNotFoundException::class)
    private fun validateUser(userId: Long?) {
        this.userRepository.findById(userId).orElseThrow { UserNotFoundException(userId) }
        val currentUserId = this.userRepository.findByUsername(SecurityContextHolder.getContext().authentication.name).get().id
        if (userId != currentUserId) {
            throw SecurityException("Cannot access another users account")
        }
    }
}
